### DISCUSSION TITLE
POST header Prefer: return=headers-only not returning LOCATION information

### DISCUSSION THREAD

**Discussion by brucer42** [PostgREST/postgrest#3451] [Q&A] [answered]
### Environment

* PostgreSQL version: 16.0 postgres
* PostgREST version: 12.0.2 postgrest/postgrest
* Operating system: macOS

### Description of issue
 
Some research indicated (to me) that this was addressed in https://github.com/PostgREST/postgrest/commit/4b46c4eff329c91dc13b26c78b266c7361ecace4#diff-b032808b94623f64b08323b5ff13143d5ce0623cd14d49b4060ad1e417a19a6c

Prefer: return=headers-only HEADER does not return location information
Prefer: return=representation does return full representation

I am using the tutorial with the todos table to reproduce this

(Steps to reproduce: Include a minimal SQL definition plus how you make the request to PostgREST and the response body)

**Comment by laurenceisla** [comment]
What is the exact request you're doing and the response you're getting? If you're seeing an error like `"permission denied for sequence..."` then you need to grant permissions to the SEQUENCE generated by `serial` as seen [in the Tutorial 1](https://postgrest.org/en/v12/tutorials/tut1.html#step-1-add-a-trusted-user).

**Comment by brucer42** [comment, marked as answer]
Purely by the tutorial, using secret 01234567012345670123456701234567, running in an emacs shell so using json_pp to print

brucer42@brucer42s-MBP postgrest % curl http://localhost:3000/todos -X POST \
     -H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoidG9kb191c2VyIn0.qrlieu5zvSLJcyDFJ_ZoWTbLkDZbVm1LAsdtPKM7J1Y"   \
     -H "Prefer: return=headers-only" \
     -H "Content-Type: application/json" \
     -d '{"task": "learn how to auth"}' | json_pp
curl http://localhost:3000/todos -X POST \
>      -H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoidG9kb191c2VyIn0.qrlieu5zvSLJcyDFJ_ZoWTbLkDZbVm1LAsdtPKM7J1Y"   \
>      -H "Prefer: return=headers-only" \
>      -H "Content-Type: application/json" \
>      -d '{"task": "learn how to auth"}' **| json_pp**
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    29    0     0  100    29      0    534 --:--:-- --:--:-- --:--:--   537
**malformed JSON string, neither array, object, number, string or atom, at character offset 0 (before "(end of string)") at /usr/bin/json_pp5.30 line 52.**
brucer42@brucer42s-MBP postgrest % curl http://localhost:3000/todos -X POST \
     -H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoidG9kb191c2VyIn0.qrlieu5zvSLJcyDFJ_ZoWTbLkDZbVm1LAsdtPKM7J1Y"   \
     -H "Prefer: return=representation" \
     -H "Content-Type: application/json" \
     -d '{"task": "learn how to auth"}' **| json_pp**
curl http://localhost:3000/todos -X POST \
>      -H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoidG9kb191c2VyIn0.qrlieu5zvSLJcyDFJ_ZoWTbLkDZbVm1LAsdtPKM7J1Y"   \
>      -H "Prefer: return=representation" \
>      -H "Content-Type: application/json" \
>      -d '{"task": "learn how to auth"}' | json_pp
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    90    0    61  100    29   4468   2124 --:--:-- --:--:-- --:--:--  6923
**[
   {
      "done" : false,
      "due" : null,
      "id" : 7,
      "task" : "learn how to auth"
   }
]**
brucer42@brucer42s-MBP postgrest %

**Reply by laurenceisla** [reply]
> malformed JSON string, neither array, object, number, string or atom, at character offset 0 (before "(end of string)") at /usr/bin/json_pp5.30 line 52.

Ah, that seems to be an output given by `json_pp` because it can't parse the content, since `Prefer: return=headers-only` returns an empty body (no content at all, only headers). Try removing `json_pp` from the  curl request when using that header and it should work.

Also add `-i` to that curl request, in order to see the `Location` header.

**Reply by brucer42** [reply]
works! thank you. See trace of it working:

brucer42@brucer42s-MBP postgrest % curl http://localhost:3000/todos -X POST \
     -H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoidG9kb191c2VyIn0.qrlieu5zvSLJcyDFJ_ZoWTbLkDZbVm1LAsdtPKM7J1Y"   \
     -H "Prefer: return=headers-only" \
     -H "Content-Type: application/json" \
     -d '{"task": "learn how to auth"}' -i
curl http://localhost:3000/todos -X POST \
<1c2VyIn0.qrlieu5zvSLJcyDFJ_ZoWTbLkDZbVm1LAsdtPKM7J1Y"   \
>      -H "Prefer: return=headers-only" \
>      -H "Content-Type: application/json" \
>      -d '{"task": "learn how to auth"}' -i
HTTP/1.1 201 Created
Transfer-Encoding: chunked
Date: Thu, 25 Apr 2024 02:15:40 GMT
Server: postgrest/12.0.2
Location: ]8;;http://localhost:3000/todos?id=eq.8\/todos?id=eq.8
]8;;\Content-Range: */*
Preference-Applied: return=headers-only

brucer42@brucer42s-MBP postgrest %
